### PR TITLE
[LLVMGPU] Add dense and paged attention runtime correctness tests

### PR DESCRIPTION
**Pull request by weimin023**
## Summary

This PR adds a CUDA end-to-end runtime correctness test for paged-KV attention as Stage 2 of [#24766](https://github.com/iree-org/iree/issues/24766).

This change also fixes and enables the shared host-side attention checker used by the existing dense-attention e2e tests. Before this change, the reference softmax was numerically incorrect and the checker generated expected results without comparing them against device output.

## Changes

- Adds CUDA paged-KV attention coverage with non-contiguous, distinct key and value page tables, `iree_linalg_ext.gather`, and `iree_linalg_ext.online_attention`.
- Makes the page-table dimension dynamic (`tensor<4x?xi64>`), matching the runtime `NUM_PAGES` use case.
- Fixes the shared stable-softmax reference and makes it compare device output with expected results.
- Refactors the paged checker to gather paged K/V into dense tensors and reuse the shared attention reference and checker.
- Passes the f16-rounded `1 / sqrt(head_dim)` scale to both the operation and host reference.

## Testing
Validated on NVIDIA RTX PRO 6000 Blackwell Server Edition

- All existing CPU dense-attention configurations.
- CUDA paged attention with 4 pages.
- The same compiled paged-attention VMFB with 2, 4, and 6 runtime pages.

Attention e2e tests use:

```text
--require_exact_results=false
--acceptable_fp_delta=0.01
```

### CODE REVIEW

File: tests/e2e/attention/CMakeLists.txt [hunk 1]
```diff
@@ -114,6 +114,27 @@ iree_generated_e2e_runner_test(
     "local"
 )
 
+iree_generated_e2e_runner_test(
```
Comment (devtbi) [inline tests/e2e/attention/CMakeLists.txt:137]: don't forget about bazel :)

Comment (weimin023) [reply tests/e2e/attention/CMakeLists.txt:137]: > don't forget about bazel :)

Hi @devtbi,
Thanks for catching this! The Bazel generated e2e macro currently skips CUDA backend/driver pairs, so a direct Bazel equivalent would be an empty test suite. Should I still add the generator and CUDA suite to a new `BUILD.bazel`, or keep this CUDA-only test CMake-only?

Comment (devtbi) [reply tests/e2e/attention/CMakeLists.txt:137]: I see, the bazel tests have been neglected in the e2e tests, let's skip that rabbit hole in this PR, has to be addressed some other time

**Review by ziereis — CHANGES_REQUESTED** [review]
i appreciate the test but i think it might be better suited here https://github.com/iree-org/iree-test-suites. Somewhere closely located with the sharktank model tests. What do you think?

CC @roberto-laudani

**Comment by weimin023** [comment]
Hi @ziereis,

Thanks for the suggestion! I agree that iree-test-suites could be a good home for model-level attention tests.

This test currently follows the existing in-tree attention e2e infrastructure: 
it uses `iree_generated_e2e_runner_test`, the existing `iree-e2e-attention-test` runner, the generator-to-MLIR calls-module flow, and the same host-side reference checking as the CPU/ROCM attention tests in `tests/e2e/attention`.

Since the existing ROCM attention e2e tests are also kept in the IREE repository, and this test is intended as a backend/runtime regression test rather than a Sharktank model test, would it make sense to keep it alongside the existing CPU/ROCM attention tests?

If the intent is to move all attention correctness tests to iree-test-suites, that may be a broader repository organization change.

**Comment by ziereis** [comment]
hmm, i see your point. Tbh to me this entire piece of code looks quite odd, just the fact that the test has to reimplement attention in C++ looks like a lot of pointless code duplication with a good amount of room for making an error in the testing implementation itself. While it could be a couple of lines to check the reference against a pytorch based implementation for example. 

Either way I see you are just extending the existing implementation and i don't want to put the burden on you moving this anywhere :) I will try to properly review this later. Thanks!

File: tests/e2e/attention/generate_e2e_paged_attention_tests.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+
+def generate_attention_mlir(output_path):
+    text = f"""\
+#executable_target_cuda = #hal.executable.target<\"cuda\", \"cuda-nvptx-fb\">
+#translation = #iree_codegen.translation_info<
+    pipeline = #iree_gpu.pipeline<VectorDistribute>
+    workgroup_size = [128, 1, 1]
+    subgroup_size = 32>
+#qk_attrs_config = #iree_gpu.lowering_config<{{
```
Comment (ziereis) [inline tests/e2e/attention/generate_e2e_paged_attention_tests.py]: a e2e test should not set the lowering config, otherwise its not e2e.

File: tests/e2e/attention/generate_e2e_paged_attention_tests.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+}}>
+#attention_lowering_config = #iree_gpu.lowering_config<{{
+    partial_reduction = [0, 0, 0, 0, 0, 8, 0],
+    workgroup = [1, 1, 1, 0, 0, 0, 0]
+}}>
+
+func.func @paged_attention(
+    %query: tensor<{BATCH}x1x1x{HEAD_DIM}xf16>,
+    %kv_storage: tensor<{POOL_PAGES}x1x{PAGE_SIZE}x{HEAD_DIM}xf16>,
+    %key_page_table: tensor<{BATCH}x{NUM_PAGES}xi64>,
```
Comment (ziereis) [inline tests/e2e/attention/generate_e2e_paged_attention_tests.py]: i think for a real model NUM_PAGES would always be a dynamic size so the test should also use a dynamic size for this.

File: tools/testing/e2e/iree-e2e-attention-test.cc [hunk 4]
```diff
@@ -443,6 +446,200 @@ class AttentionTestModuleState final {
...
+            iree_test_utils_write_element(callback_state.element_type, value,
+                                          data);
+          }
+          return iree_ok_status();
+        },
+        &callback_state, &result_view));
+    return std::move(result_view);
+  }
+
+  Status CheckPagedAttentionResults(iree_hal_device_t* device, int64_t batch,
```
Comment (ziereis) [inline tools/testing/e2e/iree-e2e-attention-test.cc:541]: this function duplicates a lot of code. It should not reimplement attention and instead just do a gather of the key and value pages and then call into the preexisting attention implementation and also use the preexisting helper functions for checking results.

**Review by ziereis — CHANGES_REQUESTED** [review]
I think there still is a lot of code duplication that can be reduced here. 

Also i noticed two more major correctness issue in the original test file that you would need to resolve as well to land this properly. 

Currently the existing reference attention implementation is wrong (https://github.com/iree-org/iree/blob/efb87ccf277e8725c06967e1442cc3504a22fc58/tools/testing/e2e/iree-e2e-attention-test.cc#L104) , it should calculate  `exp(Attention[att_idx] - max) / sum;` 

Additionally the current function for checking the attention results does not check anything:

https://github.com/iree-org/iree/blob/efb87ccf277e8725c06967e1442cc3504a22fc58/tools/testing/e2e/iree-e2e-attention-test.cc#L336

it needs validate `results->actual_contents` and `results->expected_contents`. 

If these two things are resolved, it should be possible for the implementation here to reuse a lot of the preexisting code.

**Comment by weimin023** [comment]
Hi @ziereis Thanks for the review. 

I updated the title to reflect the expanded scope: in order to reuse the shared attention test path for paged attention, this PR also fixes the existing dense-attention reference and result checker.

Does this revised scope make sense to you?

**Review by ziereis — APPROVED** [review]
Looks much better! Thanks for the cleanup :) 

Out of curiosity, did you check if the gathers actually get fused into the same dispatch as the attention op?

**Comment by weimin023** [comment]
> Looks much better! Thanks for the cleanup :)
> 
> Out of curiosity, did you check if the gathers actually get fused into the same dispatch as the attention op?

Hi @ziereis,

Yep. I checked the compilation dump and confirmed that the dispatch-creation
phase has a single `flow.dispatch.workgroups` region containing both gathers
and `online_attention`:

```mlir
%8 = flow.dispatch.workgroups[...] (...) -> tensor<4x128xf32> =
    (...) {
  ...
  %22 = iree_linalg_ext.gather
      ins(%13, %14 : tensor<8x32x128xf16>, tensor<4x?xi64>)
      outs(%21 : tensor<4x?x32x128xf16>)
      -> tensor<4x?x32x128xf16>
  %24 = iree_linalg_ext.gather
      ins(%13, %15 : tensor<8x32x128xf16>, tensor<4x?xi64>)
      outs(%21 : tensor<4x?x32x128xf16>)
      -> tensor<4x?x32x128xf16>
  ...
  %27:3 = iree_linalg_ext.online_attention
      ins(%16, %22, %24, %cst_1, %25 : ...)
      outs(...) { ... }
  ...
}
